File: chapters/35-ai-engineering-fluency-and-agent-design/go/agent_tool_call.go
package agent

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// Scenario: orchestrator executes model-proposed refund with shared idempotency semantics as Java sample.
// Demonstrates: context-bounded POST; SHA-256 key from tenant|tool|body for dedupe on retry.
// Trade-off: http.DefaultClient has no global dial timeout—production uses a tuned *http.Client transport.
func Refund(ctx context.Context, tenantID, orderID, amount string) (string, error) {
	body := fmt.Sprintf(`{"orderId":%q,"amount":%q}`, orderID, amount)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://payments.example/refunds", strings.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Idempotency-Key", idempotencyKey(tenantID, "refund", body))
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	out, err := io.ReadAll(resp.Body)
	return string(out), err
}

func idempotencyKey(tenant, tool, canonical string) string {
	sum := sha256.Sum256([]byte(tenant + "|" + tool + "|" + canonical))
	return hex.EncodeToString(sum[:])[:32]
}

File: chapters/35-ai-engineering-fluency-and-agent-design/java/AgentToolCall.java
package agent;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Duration;
import java.util.HexFormat;

// Scenario: support agent proposes a refund; orchestrator calls payments with a stable idempotency key.
// Demonstrates: SHA-256 idempotency key from tenant + tool + canonical JSON; request-level HttpClient timeout.
// Trade-off: key omits per-step UUID—good for retries; add stepId if two distinct refunds share same args.
public final class AgentToolCall {

  private static final HttpClient HTTP = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(2)).build();

  public static String refund(String tenantId, String orderId, String amount, Duration budget) throws Exception {
    String body = "{\"orderId\":\"" + orderId + "\",\"amount\":\"" + amount + "\"}";
    HttpRequest req =
        HttpRequest.newBuilder(URI.create("https://payments.example/refunds"))
            .timeout(budget)
            .header("Idempotency-Key", idempotencyKey(tenantId, "refund", body))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .build();
    HttpResponse<String> resp = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
    return resp.body();
  }

  static String idempotencyKey(String tenant, String tool, String canonicalJson) throws Exception {
    MessageDigest md = MessageDigest.getInstance("SHA-256");
    md.update((tenant + "|" + tool + "|").getBytes(StandardCharsets.UTF_8));
    md.update(canonicalJson.getBytes(StandardCharsets.UTF_8));
    return HexFormat.of().formatHex(md.digest()).substring(0, 32);
  }
}
